api/src/main/java/org/pmiops/workbench/cohorts/CohortMaterializationService.java

```diff
@@ -156,15 +158,49 @@ private ParticipantCriteria getParticipantCriteria(List<CohortStatus> statusFilt
     }
   }
 
+
+  /**
+   * Materializes a cohort.
+   * @param cohortReview {@link CohortReview} representing a manual review of participants in the cohort.
+   * @param cohortSpec JSON representing the cohort criteria.
+   * @param request {@link MaterializeCohortRequest} representing the request options
+   * @return {@link MaterializeCohortResponse} containing the results of cohort materialization
+   */
+
   public MaterializeCohortResponse materializeCohort(@Nullable CohortReview cohortReview,
-      SearchRequest searchRequest, MaterializeCohortRequest request) {
+      String cohortSpec, MaterializeCohortRequest request) {
+    SearchRequest searchRequest;
+    try {
+      searchRequest = new Gson().fromJson(cohortSpec, SearchRequest.class);
+    } catch (JsonSyntaxException e) {
+      throw new BadRequestException("Invalid cohort spec");
+    }
+    return materializeCohort(cohortReview, searchRequest, cohortSpec.hashCode(), request);
+  }
+
+  /**
+   * Materializes a cohort.
+   * @param cohortReview {@link CohortReview} representing a manual review of participants in the cohort.
+   * @param searchRequest {@link SearchRequest} representing the cohort criteria
+   * @param requestHash a number representing a stable hash of the request; used to enforce that pagination
+   *   tokens are used appropriately
+   * @param request {@link MaterializeCohortRequest} representing the request
+   * @return {@link MaterializeCohortResponse} containing the results of cohort materialization
+   */
+  @VisibleForTesting
+  MaterializeCohortResponse materializeCohort(@Nullable CohortReview cohortReview,
+      SearchRequest searchRequest, int requestHash, MaterializeCohortRequest request) {
     long offset = 0L;
     FieldSet fieldSet = request.getFieldSet();
     List<CohortStatus> statusFilter = request.getStatusFilter();
     String paginationToken = request.getPageToken();
     int pageSize = request.getPageSize();
     // TODO: add CDR version ID here
-    Object[] paginationParameters = new Object[] { searchRequest, statusFilter };
+    // We require the client to specify requestHash here instead of hashing searchRequest itself;
+    // searchRequest.hashCode() includes hashing of Enum values and is thus not stable across
+    // JVMs (see [RW-1149]).
+    Object[] paginationParameters = new Object[] { requestHash, statusFilter };
```

Comment:
`CohortStatus` is an enum, so shouldn't this fail with the same issue?

Reply:
Good catch! Fixed.